### PR TITLE
Improve section/order handling

### LINKED ISSUE #88: More intuitive section/order handling
It's annoying how you don't have to specify section order with the `@order` directive until you add the first custom section, at which point you have to explicitly list every section ID in the order. You have to cross-reference the list of default sections if you want to preserve the default sections and order.

Also, if you have redundant definitions for the same section, vimdoc should complain instead of just picking one at random.

### PR DESCRIPTION
**Pull request by dbarnett**
Fix having to explicitly declare every built-in section when you use `@order` and complain if you have redundant definitions for the same section.

Fixes #88.

Also adds a few initial tests to get the ball rolling on #29.

### CODE REVIEW

File: tests/module_tests.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  def test_duplicate_section(self):
+    plugin = module.VimPlugin('myplugin')
+    main_module = module.Module('myplugin', plugin)
+    intro = Block(vimdoc.SECTION)
+    intro.Local(name='Introduction', id='intro')
+    main_module.Merge(intro)
+    intro2 = Block(vimdoc.SECTION)
+    intro2.Local(name='Intro', id='intro')
+    with self.assertRaises(error.DuplicateSection) as cm:
```
Comment (malcolmr) [inline tests/module_tests.py]: why `as cm`?  (just `with self.assertRaises(...):`?)

File: vimdoc/error.py [hunk 2]
```diff
@@ -116,6 +116,18 @@ def __init__(self, section):
         'Section {} never defined.'.format(section))
 
 
+class DuplicateSection(BadStructure):
+  def __init__(self, section):
+    super(DuplicateSection, self).__init__(
```
Comment (malcolmr) [inline vimdoc/error.py]: nit: since (IIRC) Vimdoc now requires Python 3, this can just be `super().__init__(`.

File: tests/module_tests.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import unittest
+
+import vimdoc
+from vimdoc.block import Block
+from vimdoc import error
+from vimdoc import module
+
+class TestVimPlugin(unittest.TestCase):
```
Comment (malcolmr) [inline tests/module_tests.py]: `TestModule`, perhaps? That's what we're testing here, isn't it?

File: vimdoc/module.py [hunk 4]
```diff
@@ -141,25 +142,35 @@ def Close(self):
...
+        'autocmds',
+        'settings',
+        'dicts',
+        'functions',
+        'exceptions',
+        'mappings']:
+      if builtin in self.sections or builtin in self.backmatters:
+        if builtin not in order:
+          builtins_pre.append(builtin)
+    order = builtins_pre + order
```
Comment (malcolmr) [inline vimdoc/module.py]: Isn't this going to run into trouble if `@order` _does_ contain one of the builtins? (i.e. it'll end up in `order` twice).

Some tests for this part would be nice, if possible.

**Comment by malcolmr** [comment]
Mostly looks reasonable, but see above.

File: vimdoc/module.py (hunk 4 above)
Comment (dbarnett) [inline vimdoc/module.py]: Don't think so. It only adds if it's not already there (`if builtin not in order:`).

I struggled with making the ordering smarter so it could stay closer to the default ordering. For instance, if you had:
`@order intro custom1 custom2 config about` and also defined sections for commands and functions, this implementation would give you
  `commands functions intro custom1 custom2 config about`
instead of
  `intro custom1 custom2 config commands functions about`
I had trouble even coming up with the words to describe the better behavior clearly.

If that order feels more intuitive, I can try to update it before submitting.

File: vimdoc/module.py [hunk 5]
```diff
@@ -60,11 +60,17 @@ def Merge(self, block, namespace=None):
       # Overwrite existing section if it's a default.
       if block_id not in self.sections or self.sections[block_id].IsDefault():
```
Comment (malcolmr) [inline vimdoc/module.py]: This might be simpler if we just invert that check, unless there's some reason we need to be able to support overwriting a default section with another default section?

```python
# Allow overwriting an existing section if it's a default.
if block_id in self.sections and not self.sections[block_id].IsDefault():
  raise error.DuplicateSection(block_id)
self.sections[block_id] = block
```

File: vimdoc/module.py (hunk 4 above)
Comment (malcolmr) [inline vimdoc/module.py]: I think it's reasonable to stick the user-specified parts in the middle of the list for now, though perhaps it'd be worth taking a look at how this is used in practice - I can certainly see the attraction of the more complicated ordering.

However, I'm getting a little bit confused here. `backmatters` must have a subset of the keys of `sections`, I think, otherwise we fail just below, at L163-165.

However, we then set `known = set(self.sections) | set(self.backmatters)`, which must surely just be the same as `set(self.sections)`, unless I'm missing something?

Either way, if we pull up L163-166 to before this, we can perhaps write this more simply (WDYT?):

```python
DEFAULT_ORDER_PRE = [
  'intro',
  'config',
  'commands',
  'autocmds',
  'settings',
  'dicts',
  'functions',
  'exceptions',
  'mappings']
DEFAULT_ORDER_POST = [
  'about']
```
```python
for backmatter in self.backmatters:
  if backmatter not in self.sections:
    raise error.NoSuchSection(backmatter)
known = set(self.sections)  # I think?

explicit_order = self.order or []
order_pre = [section for section in DEFAULT_ORDER_PRE
             if section in known and section not in explicit_order]
order_post = [section for section in DEFAULT_ORDER_POST
              if section in known and section not in explicit_order]
order = order_pre + explicit_order + order_post
```

(Incidentally, what indent are we aiming for, PEP8 or Google? We seem to use both.)

File: vimdoc/module.py (hunk 5 above)
Comment (dbarnett) [inline vimdoc/module.py]: I actually tried that at first, but there's a third case that's missing (default/default shouldn't throw error). Just played with a few variations on inverting the check again but didn't find any better version.

Comment (malcolmr) [inline vimdoc/module.py]: I'm surprised we _are_ trying to overwrite non-defaults with defaults, but if that's what we need to do, then this is certainly reasonable.

**Comment by malcolmr** [comment]
LGTM whichever way you decide to go on the above, btw.

File: vimdoc/module.py (hunk 5 above)
Comment (dbarnett) [inline vimdoc/module.py]: Yeah, it's a matter of traversal order. Special flags like `plugin[foo]` can have a default from one file (plugin/foo.vim) and explicit value declared in another, so we don't know which definition will come last and try to "overwrite" the other.

**Comment by dbarnett** [comment]
BTW, I'm still updating and addressing the other comments.

I think I do want to get the implicit order into the "natural order" that makes intuitive sense to me, but I'm still having trouble describing the algorithm in plain English. Anyway, I'm adding some tests for the ordering behavior since it's clearly non-trivial. =)

File: vimdoc/error.py (hunk 2 above)
Comment (dbarnett) [inline vimdoc/error.py]: Not worth going out of our way to use the new convention, though.

File: tests/module_tests.py (hunk 3 above)
Comment (dbarnett) [inline tests/module_tests.py]: Went with `TestVimModule` since `TestModule` sounded like a generic "this is a module of tests" to me.

File: vimdoc/module.py (hunk 4 above)
Comment (dbarnett) [inline vimdoc/module.py]: You're right the `backmatter` checks were unnecessary. The code looks a lot different now for the fancier ordering, but I applied several of your suggestions first.

**Comment by dbarnett** [comment]
PTAL. Improved the default ordering and added more tests for it.

File: vimdoc/module.py [hunk 6]
```diff
@@ -200,6 +190,46 @@ def Chunks(self):
...
+      'autocmds',
+      'settings',
+      'dicts',
+      'functions',
+      'exceptions',
+      'mappings']
+    # Add any undeclared sections before custom sections, except 'about' which
+    # comes at the end by default.
+    section_insertion_idx = 0
+    order = order[:]
```
Comment (malcolmr) [inline vimdoc/module.py]: Any reason not to take a copy at the point of assignment above?

i.e.
```python
order = explicit_order[:] or []
```

**Comment by malcolmr** [comment]
LGTM, thanks (and sorry for the delay).

File: vimdoc/module.py (hunk 6 above)
Comment (dbarnett) [inline vimdoc/module.py]: Yeah, I tried that version first, but it fails for `None`.
